docs/pipeline/triggering/PERFORMANCE_TRIGGERS.md

```diff
@@ -0,0 +1,180 @@
+    PerformanceTriggerCriterion <|-- NumberAvoidableMisclassificationCriterion
+    NumberAvoidableMisclassificationCriterion <|-- StaticNumberAvoidableMisclassificationCriterion
+
+```
+
+### `PerformanceDecisionPolicy` Hierarchy
+
+The `PerformanceDecisionPolicy` is an abstract base class that defines the decision-making process for performance triggers. It is extended by specific policies that define how decisions are made:
+
+1. **StaticPerformanceThresholdDecisionPolicy**: Triggers when the performance metric falls below a static threshold.
+2. **DynamicPerformanceThresholdDecisionPolicy**: Triggers when the performance metric deviates beyond a dynamic threshold, based on a rolling average.
+3. **StaticNumberAvoidableMisclassificationDecisionPolicy**: Triggers when the number of avoidable misclassifications exceeds a predefined static threshold.
+
+<details>
+<summary><b>What are avoidable misclassification?</b></summary>
+
+Avoidable misclassifications are the misclassifications that could have been avoided if the model had triggered earlier. They can be seen as the lost benefits of not triggering. Every evaluation point we estimate the avoidable misclassifications by comparing the expected performance (the performance of the model if we had triggered) with the actual performance (the performance of the model since the last trigger). Avoidable misclassifications are then derived from the difference between the expected and actual accuracy and the estimated query density.
+
+</details>
+
+<details>
+<summary><b>Hindsight vs. Forecasting mode</b></summary>
+All decision work on estimates of e.g. the expected performance, next actual/observed performance (if no trigger is executed now) and the estimated query density. The estimates can be made in two modes: `hindsight` and `forecasting`.
+
+In hindsight mode, decisions are based solely on the current observed performance metrics. This mode checks if the most recent performance falls below a predefined threshold or deviates significantly from the expected performance, based on past evaluations. It is a retrospective approach, relying on already observed data to determine if a trigger should be invoked. Triggering happens post-factum when the performance has already degraded / a threshold has been crossed.
+
+On the other hand, forecasting mode looks forward and incorporates predictions of future performance into the decision-making process. This mode not only evaluates the current performance but also forecasts future performance and data density to anticipate potential issues before they occur. For example, in the static number avoidable misclassification policy, forecasting mode predicts whether the cumulative number of avoidable misclassifications will exceed the threshold before the next evaluation. By doing so, it aims to prevent performance degradation by triggering preemptively, rather than waiting for the problem to manifest.
```

Comment:
I guess what is unclear to me is a) the role of data density here (what is it, what is it needed for) and 2) if I understand correctly, you predict the number of avoidable missclassifications? If this is the case, maybe we should make clear that's its potentially difficult to predict sudden drift? because i was wondering how e.g. the drift in yearbook would have been predicted

Reply:
1) see other threads; I'll add more background info for this.
2) That's true; it's not possible to forecast sudden drift with this approach. Not sure if it's possible with any approach at all. The whole idea is inspired by slowly/gradually drifting datasets like we see in arxiv. In those instances, we would never see sudden drift but build a cumulative "regret" metric (which is the number of misclassifications here). For the forecasted variant, we use the same expected performance (that is derived from the series of evaluations directly after triggers) but use a forecasted accuracy instead of the last accuracy measurement on the last batch. The forecasted accuracy is currently only a linear extrapolation through a ridge regression model (more sophisticated time series forecasting methods are possible obviously)

Reply:
That makes sense. I wonder if something like a cumulative regret could also be built for drift (or if in drift, one could check out the drift w.r.t. some older data) but yeah thanks for clarifying the density.

Reply:
For the number of misclassifications this cumulative approach feels rather natural. For drift I'm not sure if it makes sense to remember drift areas that have already been passed. I thought of this in the `CostTrigger` domain, where you have a regret metric (maybe performance, maybe drift, maybe data amount based incorporation latency) and a cost estimate (training time). However, for drift this would almost be impossible to configure for a user. We need this conversion rate hyperparameter from the user. And I don't see any semantic connection / heuristic a user could employ to derive a relation between an arbitrary drift distance sum and training time.